Runtime/Hosting/NativeHost.cs

```diff
@@ -0,0 +1,172 @@
+                managedHostTypeAssemblyQualifiedName,
+                nameof(InitializeModule),
+                delegateType: -1 /* UNMANAGEDCALLERSONLY_METHOD */,
+                reserved: default,
+                out nint initializeModulePointer);
+            CheckStatus(status, "Failed to load managed host assembly.");
+
+            // Invoke the managed host initialize method.
+            // (It will define some properties on the exports object passed in.)
+            napi_register_module_v1 initializeModule =
+                Marshal.GetDelegateForFunctionPointer<napi_register_module_v1>(
+                    initializeModulePointer);
+            initializeModule(env, exports);
+        }
+        catch
+        {
+            hostfxr_close(_hostContextHandle);
+            _hostContextHandle = default;
+            throw;
+        }
+    }
+
+    public void Dispose()
+    {
+        // Close the CLR host context handle, if it's still open.
+        if (_hostContextHandle != default)
+        {
+            hostfxr_status status = hostfxr_close(_hostContextHandle);
+            _hostContextHandle = default;
+            CheckStatus(status, "Failed to dispose CLR host.");
+        }
+    }
+
+    private static void CheckStatus(hostfxr_status status, string message)
+    {
+        if (status != hostfxr_status.Success &&
+            status != hostfxr_status.Success_HostAlreadyInitialized)
+        {
+            throw new Exception(Enum.IsDefined(status) ?
+                $"{message} Status: {status}" : $"{message} HRESULT: 0x{(uint)status:x8}");
+        }
+    }
+
+    private static unsafe string GetCurrentModuleFilePath()
+    {
+        // TODO: Use dladdr() on non-Windows systems to get the current module file path.
+        // Unfortunately Assembly.Location/Codebase doesn't work for an AOT compiled library.
+
+        delegate* unmanaged[Cdecl]<napi_env, napi_value, napi_value> functionInModule =
+            &InitializeModule;
+
+        const uint GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS = 0x00000004;
+        if (GetModuleHandleExW(
+            GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS,
+            (nint)functionInModule,
+            out nint moduleHandle) == 0)
+        {
+            throw new Exception("Failed to get current module handle.");
+        }
+
+        StringBuilder filePathBuilder = new(255);
+        uint filePathLength = GetModuleFileName(
+            moduleHandle, filePathBuilder, filePathBuilder.Capacity);
+        if (filePathLength == 0)
+        {
+            throw new Exception("Failed to get current module file path.");
+        }
+
+        return filePathBuilder.ToString();
+    }
+
+    [LibraryImport("kernel32", SetLastError = true)]
+    private static partial uint GetModuleHandleExW(
+        uint flags,
+        nint moduleNameOrAddress,
+        out nint moduleHandle);
+
+    [DllImport("kernel32", SetLastError = true, CharSet = CharSet.Unicode)]
```

Comment:
>DllImport [](http://example.com/codeflow?start=5&length=9)

I am just curious why LibraryImport did not work here.

Reply:
`[LibraryImport]` doesn't support use of `StringBuilder` for getting string output.

In the latest revision I've changed this to `[LibraryImport]` using a stack-allocated `char*` instead.